concordium-consensus/src/Concordium/GlobalState/LMDB/Helpers.hs

```diff
@@ -132,67 +138,67 @@ acquireRead RWLock{..} = mask_ go
             go
       st@(ReadLocked n waitingWriters)
           | waitingWriters == 0 ->
-            -- No waiting writers, add another reader.
+            -- No waiting writers, add another reader and succeed.
             putMVar rwlState $! ReadLocked (n + 1) 0
           | otherwise -> do
-              -- Some readers hold the lock, but there are waiting writers.
-              -- We do nothing and wait until there are no more readers and attempt again.
-              -- At that point we are likely to end up in WriteLocked state.
+              -- Some readers hold the lock, but there are waiting writers. We
+              -- do nothing and wait until there are no more readers and attempt
+              -- again. At that point we are likely to end up in WriteLocked
+              -- state. If not, we do another loop relying on fairness of
+              -- waitQSem to make progress.
               putMVar rwlState st
-              readMVar rwlReadLock
+              waitQSem rwlReadLock
               go
       lockState@(WriteLocked _) -> do
         -- There is an active writer. Do nothing and wait until the writer is done.
         putMVar rwlState lockState
-        readMVar rwlWriteLock
+        waitQSem rwlWriteLock
```

Comment:
I think this can also lead to a deadlock if this semaphore is not signalled.

Reply:
Yes, this is reverted to use MVars as it was originally.